sdk/java/runtime/main.go

```diff
@@ -0,0 +1,312 @@
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"java-sdk/internal/dagger"
+
+	"github.com/iancoleman/strcase"
+)
+
+const (
+	MavenImage  = "maven:3.9.9-eclipse-temurin-23-alpine"
+	MavenDigest = "sha256:77fe6f79f868484d85679bfaa121b34177e3fedf570086ea6babbd6db2223e89"
+	JavaImage   = "eclipse-temurin:23-jre-alpine"
+	JavaDigest  = "sha256:bd8e2c8c19bcadbaa8c6a128051a22384c6f7cfe5fa520cb663fe21fff96f084"
+
+	ModSourceDirPath = "/src"
+	ModDirPath       = "/opt/module"
+	GenPath          = "sdk"
+)
+
+type JavaSdk struct {
+	SDKSourceDir  *dagger.Directory
+	RequiredPaths []string
+	moduleConfig  moduleConfig
+}
+
+type moduleConfig struct {
+	name    string
+	subPath string
+}
+
+func (c *moduleConfig) modulePath() string {
+	return filepath.Join(ModSourceDirPath, c.subPath)
+}
+
+func (c *moduleConfig) sdkPath() string {
+	return filepath.Join(c.modulePath(), GenPath)
+}
+
+func New(
+	// Directory with the Java SDK source code.
+	// dagger-java-samples is not necessary to build, but as it's referenced in the root pom.xml maven
+	// will check if it's there. So we keep it, until a better solution (like to extract the samples somewhere else)
+	// +optional
+	// +defaultPath="/sdk/java"
+	// +ignore=["**", "!dagger-codegen-maven-plugin/", "!dagger-java-annotation-processor/", "!dagger-java-sdk/", "!dagger-java-samples", "!LICENSE", "!README.md", "!pom.xml"]
+	sdkSourceDir *dagger.Directory,
+) (*JavaSdk, error) {
+	if sdkSourceDir == nil {
+		return nil, fmt.Errorf("sdk source directory not provided")
+	}
+	return &JavaSdk{
+		RequiredPaths: []string{},
+		SDKSourceDir:  sdkSourceDir,
+	}, nil
+}
+
+func (m *JavaSdk) Codegen(
+	ctx context.Context,
+	modSource *dagger.ModuleSource,
+	introspectionJSON *dagger.File,
+) (*dagger.GeneratedCode, error) {
+	if err := m.setModuleConfig(ctx, modSource); err != nil {
+		return nil, err
+	}
+	mvnCtr := m.codegenBase(ctx, modSource, introspectionJSON)
+
+	return dag.
+		GeneratedCode(dag.Directory().WithDirectory("/", m.generateCode(ctx, mvnCtr, introspectionJSON))).
+		WithVCSGeneratedPaths([]string{
+			"target/generated-sources/dagger/**",
+			"target/generated-sources/entrypoint/**",
+			"target/generated-sources/sdk/**",
+		}).
+		WithVCSIgnoredPaths([]string{
+			"target",
+		}), nil
+}
+
+// codegenBase takes the user module code, add the generated SDK dependencies
+// if the user module code is empty, creates a default module content based on the template from the SDK
+// The generated container will *not* contain the SDK source code, but only the packages built from the SDK
+func (m *JavaSdk) codegenBase(
+	ctx context.Context,
+	modSource *dagger.ModuleSource,
+	introspectionJSON *dagger.File,
+) *dagger.Container {
+	return m.
+		// We need maven to build the user module
+		mvnContainer(ctx).
+		// Mount the maven folder where we installed all the generated dependencies
+		WithMountedDirectory("/root/.m2", m.buildJavaDependencies(ctx, introspectionJSON).Directory("/root/.m2")).
```

Comment:
Curious, is `$HOME/.m2` a maven convention?

Reply:
Yes it is `$HOME/.m2` contains the maven configuration files for the local user and maven stores all downloaded dependencies in the `$HOME/.m2/repository` directory